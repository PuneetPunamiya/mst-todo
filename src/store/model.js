import { types, unprotect } from "mobx-state-tree"

const Todo = types.model({
  task: types.string,
  done: types.optional(types.boolean, false)
})
.actions(self => ({

  toggle() {
      self.done = !self.done
  }
}))

export const TodoStore = types.model({
  todos: types.array(Todo)
})

const s = TodoStore.create({ todos: [] })
unprotect(s) // needed to allow modifying outside of an action
s.todos.push({ task: "Grab coffee" })
// console.log(s.todos[0]) // prints: "Grab coffee"