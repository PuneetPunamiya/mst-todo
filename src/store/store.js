import { types } from 'mobx-state-tree';

const catalog = types.model({
  id: types.identifier,
  name: types.optional(types.string, ''),
  type: types.optional(types.string, ''),
});

const Todo = types
  .model({
    id: types.string,
    name: types.optional(types.string, ''),
    version: types.optional(types.string, ''),
    catalog: types.maybe(types.reference(catalog)),
    done: types.optional(types.boolean, false),
  })
  .actions((self) => ({
    setName(newName) {
      self.name = newName;
    },

    toggle() {
      self.done = !self.done;
    },
  }));

const User = types.model({
  name: types.optional(types.string, ''),
});

export const RootStore = types
  .model({
    users: types.map(User),
    todos: types.map(Todo),
    catalog: types.map(catalog),
  })
  .actions((self) => ({
    addTodo(item) {
      self.todos.set(item.id, {
        id: item.id,
        name: item.name,
        version: item.version,
        catalog: item.catalog,
        done: item.done,
      });
    },
    addCatalog(item) {
      self.catalog.put({
        id: item.id,
        name: item.name,
        type: item.type,
      });
    },
  }));

const store = RootStore.create({});
