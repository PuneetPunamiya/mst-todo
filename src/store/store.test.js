import { RootStore } from "./store";
import { getSnapshot } from "mobx-state-tree";

it("creates a store", () => {
  const store = RootStore.create({});
  store.addTodo(1, "Eat a cake");

  store.addTodo(1, "Eat a cake");
  expect(store.todos.size).toBe(1);
  store.todos.get(1).toggle();
  expect(store.todos.get(1).done).toBe(true);
});

it("creates a store to match snapshot", () => {
  const store = RootStore.create({});
  store.addTodo(1, "Eat a cake");

  store.addTodo(1, "Eat a cake");
  expect(store.todos.size).toBe(1);
  store.todos.get(1).toggle();
  expect(store).toMatchSnapshot();
});
