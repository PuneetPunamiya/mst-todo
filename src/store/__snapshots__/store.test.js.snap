// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a store to match snapshot 1`] = `
Object {
  "todos": Object {
    "1": Object {
      "done": true,
      "name": "Eat a cake",
    },
  },
  "users": Object {},
}
`;
